cirq-core/cirq/sim/clifford/stabilizer_state_ch_form.py

```diff
@@ -98,6 +98,62 @@ def copy(self) -> 'cirq.StabilizerStateChForm':
 
         return copy
 
+    def join(self, other: 'cirq.StabilizerStateChForm') -> 'cirq.StabilizerStateChForm':
+        n = self.n + other.n
+        copy = StabilizerStateChForm(n)
+
+        def diag(a, b):
+            a = np.pad(a, [0, other.n])
+            b = np.pad(b, [self.n, 0])
+            return a + b
+
+        copy.G = diag(self.G, other.G)
+        copy.F = diag(self.F, other.F)
+        copy.M = diag(self.M, other.M)
+        copy.gamma = np.concatenate([self.gamma, other.gamma])
+        copy.v = np.concatenate([self.v, other.v])
+        copy.s = np.concatenate([self.s, other.s])
+        copy.omega = self.omega * other.omega
+        return copy
+
+    def extract(
+        self, axes: Sequence[int]
+    ) -> Tuple['cirq.StabilizerStateChForm', 'cirq.StabilizerStateChForm']:
+        n = len(axes)
+        remaining_axes = [i for i in range(self.n) if i not in axes]
+        extracted = StabilizerStateChForm(n)
+        indices = list(axes) + remaining_axes
+        G = self.G[indices][:, indices]
+        F = self.F[indices][:, indices]
+        M = self.M[indices][:, indices]
+        extracted.G = G[:n, :n]
+        extracted.F = F[:n, :n]
+        extracted.M = M[:n, :n]
+        extracted.gamma = self.gamma[axes]
+        extracted.v = self.v[axes]
+        extracted.s = self.s[axes]
+        extracted.omega = self.omega ** (len(axes) / self.n)
+        remainder = StabilizerStateChForm(len(remaining_axes))
+        remainder.G = G[n:, n:]
+        remainder.F = F[n:, n:]
+        remainder.M = M[n:, n:]
+        remainder.gamma = self.gamma[remaining_axes]
+        remainder.v = self.v[remaining_axes]
+        remainder.s = self.s[remaining_axes]
+        remainder.omega = self.omega ** (len(remaining_axes) / self.n)
```

Comment:
I think the `omega` may not be right here. Unit tests pass but the hidden_linear_shift.ipynb fails and it uses CliffordSimulator. If I change the notebook to use sparse simulator, the hidden shift works again. I can't decipher what `omega` represents, so I anticipate this is the problem. @95-martin-orion who would be the best person to look at this?

For context this is the function that gets called after measuring an axis to split the state space into two disjoint spaces. The `axes` argument is always the axis that was measured. Hopefully such an operation is well-defined in terms of the CH representation, as it provides a linear speedup.

Reply:
Given #2423 it seems the experts are @Strilanc and @smitsanghavi. Can you have a look? Hopefully there's a well defined (and not too expensive) way to get the omega values.

Reply:
Hey Dax, `omega` represents the global phase. I'm not 100% what's the intention here. When you say that this function splits "the state space into two disjoint spaces", are we creating a mixture? 
If it's just about representing the full state as a tensor of two disjoint states, the current implementation looks reasonable to me (IIUC, this smears the global phase is uniformly across the qubits such that it multiplies to original `omega` in `join`).

Reply:
Thanks, the intention here is if you have two subsets of qubits that you can be sure are unentangled (e.g. you just measured one), then we split them into two groups and simulate them independently. And you do the opposite when you have two separate groups you're simulating and there's a gate that entangles them. This allows exponential speedup on state vector and density matrix simulators, and should be up to a linear speedup on CH-Form AFAICT.

I've tried a few things but nothing has worked 100% yet. If I can't figure it out, maybe we can postpone Clifford since the improvement isn't as dramatic as the other two.

Reply:
Hmm, I've tried everything I can think of. Resetting extracted.omega =1, various "smearing" approaches, even having a reference cell that gets shared across all the disjoint sets (reasonable if it's a global shift). Nothing works. I'm thinking the CH form just isn't conducive to distributing in this fashion, that it's designed to work all together as a whole. All the splitting and joining might require maintaining local phase shift per qubit, but that's way outside the scope of this PR.

I'll think about it a little more but pretty sure at this point I'll just remove support for this from the Clifford simulator and add a work item to investigate later on. A bit of a bummer since it feels so close, but oh well.

Reply:
If even having a shared `omega` isn't working, the issue might be somewhere else. 

Have you tried modifying the failing notebook to have it compare the state vectors using `protocols.equal_up_to_global_phase` instead? This would confirm if the issue is with the global phase or some other param.

Reply:
The notebook fails higher level, in finding the hidden shift, not a direct check of a quantum state.

I think the thing is, even with a shared omega, CH form algorithm no longer does the same operations it would do on the various matrices; it skips any operations it would have done on external qubit sets. I haven't fully gone through the algo, but it's conceivable that CH form is specially designed to work in only the way it does, and splitting / re-merging just isn't compatible with that form. Maybe we could invent a new form that's similar but is compatible with splitting and merging. I'll have to look into it, but for now I think I'll put it as a TODO.